FlowCrypt/Functionality/Services/EKMService.swift

```diff
@@ -0,0 +1,188 @@
+//
+//  EKMService.swift
+//  FlowCrypt
+//
+//  Created by Ioan Moldovan on 4/28/22
+//  Copyright © 2017-present FlowCrypt a. s. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+protocol EKMServiceType {
+    func refreshKeysFromEKMIfNeeded(in viewController: UIViewController)
+}
+
+enum RefreshKeyError: Error {
+    case cancelPassPhrase
+}
+
+final class EKMService: EKMServiceType {
+
+    let appContext: AppContextWithUser
+    var askedUserPassPhrase = false
+    let keyMethods: KeyMethodsType
+
+    init(appContext: AppContextWithUser) {
+        self.appContext = appContext
+        self.keyMethods = KeyMethods()
+    }
+
+    func refreshKeysFromEKMIfNeeded(in viewController: UIViewController) {
+        Task {
+            do {
+                self.askedUserPassPhrase = false
+                let configuration = try await appContext.clientConfigurationService.configuration
+                guard configuration.checkUsesEKM() == .usesEKM else {
+                    return
+                }
+                let emailKeyManagerApi = EmailKeyManagerApi(clientConfiguration: configuration)
+                let idToken = try await IdTokenUtils.getIdToken(userEmail: appContext.user.email)
+                let keyDetails = try await emailKeyManagerApi.getPrivateKeys(idToken: idToken)
+                let localKeys = try appContext.encryptedStorage.getKeypairs(by: appContext.user.email)
+
+                var passPhrase = try fetchCurrentPassphrase()
+                let keysToUpdate = try findKeysToUpdate(from: keyDetails, localKeys: localKeys)
+
+                if keysToUpdate.isEmpty {
+                    return
+                }
+
+                for keyDetail in keysToUpdate {
+                    passPhrase = try await saveKeyToLocal(
+                        in: viewController,
+                        context: appContext,
+                        keyDetail: keyDetail,
+                        passPhrase: passPhrase
+                    )
+                }
+                await viewController.showToast("refresh_key_success".localized)
+            } catch {
+                if error is ApiError {
+                    return
+                }
+                // Do not display error alert when no keys are returned
+                if let ekmError = error as? EmailKeyManagerApiError, ekmError == .noKeys {
+                    return
+                }
```

Comment:
We should change this in the future in another PR. `emailKeyManagerApi.getPrivateKeys` should not throw when the list is empty, handling that should be the job of the code that called it. That would mean we could also remove this check.

Reply:
You can, but the scope is getting quite large in this PR

Reply:
ok, already changed though